velox/expression/SignatureBinder.cpp

```diff
@@ -229,80 +307,53 @@ bool SignatureBinderBase::checkOrSetIntegerParameter(
-  return tryBind(typeSignature, actualType, true, coercion);
-}
-
-bool SignatureBinderBase::tryBind(
-    const exec::TypeSignature& typeSignature,
-    const TypePtr& actualType) {
-  Coercion coercion;
-  return tryBind(typeSignature, actualType, false, coercion);
-}
-
 bool SignatureBinderBase::tryBind(
     const exec::TypeSignature& typeSignature,
     const TypePtr& actualType,
-    bool allowCoercion,
-    Coercion& coercion) {
-  coercion.reset();
+    Coercion* coercion) {
+  if (!actualType) {
+    return false;
+  }
+
   if (isAny(typeSignature)) {
     return true;
   }
 
   const auto& baseName = typeSignature.baseName();
 
-  if (variables().count(baseName)) {
-    // Variables cannot have further parameters.
+  if (auto varIt = variables().find(baseName); varIt != variables().end()) {
     VELOX_CHECK(
         typeSignature.parameters().empty(),
         "Variables with parameters are not supported");
-    auto& variable = variables().at(baseName);
-    VELOX_CHECK(variable.isTypeParameter(), "Not expecting integer variable");
-
-    if (typeVariablesBindings_.count(baseName)) {
-      // If the the variable type is already mapped to a concrete type, make
-      // sure the mapped type is equivalent to the actual type.
-      return typeVariablesBindings_[baseName]->equivalent(*actualType);
-    }
+    VELOX_CHECK(varIt->second.isTypeParameter());
 
-    if (actualType->isUnKnown() && variable.knownTypesOnly()) {
-      return false;
-    }
-
-    if (variable.orderableTypesOnly() && !actualType->isOrderable()) {
-      return false;
+    auto& varType = typeVariablesBindings_[varIt->second.name()];
+    if (!varType) {
+      if (!checkSignatureProperties(varIt->second, actualType)) {
+        return false;
+      }
+      varType = actualType;
     }
-
-    if (variable.comparableTypesOnly() && !actualType->isComparable()) {
-      return false;
+    if (!coercion) {
+      return varType->equivalent(*actualType);
     }
-
-    typeVariablesBindings_[baseName] = actualType;
-    return true;
+    *coercion = TypeCoercer::coercible(actualType, varType);
+    return static_cast<bool>(*coercion);
   }
 
+  const auto& params = typeSignature.parameters();
+
   // Type is not a variable.
   auto typeName = boost::algorithm::to_upper_copy(baseName);
-  auto actualTypeName =
-      boost::algorithm::to_upper_copy(std::string(actualType->name()));
-
-  if (typeName != actualTypeName) {
-    if (allowCoercion) {
-      if (auto availableCoercion =
-              TypeCoercer::coerceTypeBase(actualType, typeName)) {
-        coercion = availableCoercion.value();
-        return true;
-      }
+  if (!boost::algorithm::iequals(typeName, actualType->name())) {
+    if (!coercion || !params.empty()) {
+      return false;
     }
-    return false;
+    // TODO: Handle coercion for complex types with parameters.
+    *coercion = TypeCoercer::coerceTypeBase(actualType, typeName);
+    return static_cast<bool>(*coercion);
```

Comment:
> static_cast<bool>(*coercion);

I feel code like this is hard to read an understand. Is there a way to simplify?

Reply:
I followed second option, because I think you prefer it, see https://github.com/facebookincubator/velox/pull/15734/changes/861117750d189ac36413499955e1c28d07e1fb42